src/System.Text.Primitives/System/Text/Encoding/Utf8/Utf8Encoder.cs

```diff
@@ -290,53 +300,119 @@ internal static bool TryComputeEncodedBytes(ReadOnlySpan<uint> utf32, out int by
+                non_ascii:
+                for (; i < inputCharacters; i++)
+                {
+                    codePoint = *(utf16CharRef + i);
+                    if (codePoint <= Utf8OneByteLastCodePoint)
+                    {
+                        goto ascii;
+                    }
+                    else if (codePoint <= Utf8TwoBytesLastCodePoint)
+                    {
+                        if (bytesWrittenTemp + 2 > outputBytesAvailable)
+                        {
+                            goto need_more;
+                        }
+                        *(utf8ByteRef + bytesWrittenTemp) = (byte)(((codePoint >> 6) & b0001_1111U) | b1100_0000U);
+                        bytesWrittenTemp++;
+                        *(utf8ByteRef + bytesWrittenTemp) = (byte)(((codePoint) & b0011_1111U) | b1000_0000U);
+                        bytesWrittenTemp++;
+                    }
+                    else if (codePoint >= Utf16HighSurrogateFirstCodePoint && codePoint <= Utf16HighSurrogateLastCodePoint)
+                    {
+                        if (bytesWrittenTemp + 4 > outputBytesAvailable)
+                        {
+                            goto need_more;
+                        }
+
+                        if (++i >= inputCharacters)
+                            throw new ArgumentException("Invalid surrogate pair.", nameof(utf16));
+
+                        var lowSurrogate = *(utf16CharRef + i);
+
+                        if (lowSurrogate < Utf16LowSurrogateFirstCodePoint || lowSurrogate > Utf16LowSurrogateLastCodePoint)
+                            throw new ArgumentException("Invalid surrogate pair.", nameof(utf16));
+
+
+                        int surrogateCalculation;
+                        unchecked
+                        {
+                            surrogateCalculation = (((codePoint - Utf16HighSurrogateFirstCodePoint) << 10)
+                            | (lowSurrogate - Utf16LowSurrogateFirstCodePoint)) + 0x10000;
+                        }
+
+                        *(utf8ByteRef + bytesWrittenTemp) = (byte)(((surrogateCalculation >> 18) & b0000_0111U) | b1111_0000U);
+                        bytesWrittenTemp++;
+                        *(utf8ByteRef + bytesWrittenTemp) = (byte)(((surrogateCalculation >> 12) & b0011_1111U) | b1000_0000U);
+                        bytesWrittenTemp++;
+                        *(utf8ByteRef + bytesWrittenTemp) = (byte)(((surrogateCalculation >> 6) & b0011_1111U) | b1000_0000U);
+                        bytesWrittenTemp++;
+                        *(utf8ByteRef + bytesWrittenTemp) = (byte)(((surrogateCalculation) & b0011_1111U) | b1000_0000U);
+                        bytesWrittenTemp++;
+                    }
+                    else if (codePoint <= Utf8ThreeBytesLastCodePoint)
                     {
-                        codePoint = (uint)ch;
-                        consumed = 1;
+                        if (bytesWrittenTemp + 3 > outputBytesAvailable)
```

Comment:
You need to write the code such that it handles the boundary conditions well. In this case, consider what will happen for: 

```
outputBytesAvailable = 0x7FFFFFFE;
bytesWrittenTemp = 0x7FFFFFFD;

if (bytesWrittenTemp + 3 > outputBytesAvailable) // bytesWrittenTemp + 3 will overflow will give you -MaxInt, and this condition will be incorrectly false
```

In this case, you can fix it by rewriting the condition as:
```
                  if (bytesWrittenTemp > outputBytesAvailable - 3)
```

Also, you may want to take some of the security trainings about security bugs caused by integer overflows… .